### PR TITLE
[AIRFLOW-2566] Change backfill to rerun failed tasks

### PR DESCRIPTION
**Pull request by feng-tao**
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2566
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
1. remove some debug messages for backfill
2. change the backfill to only rerun the failed tasks instead of throwing exceptions.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by feng-tao** [comment]
will wait for https://github.com/apache/incubator-airflow/pull/3465

**Comment by feng-tao** [comment]
This build(https://travis-ci.org/apache/incubator-airflow/builds/389515754) shows all the tests pass and flake8 fail to run.

**Comment by feng-tao** [comment]
PTAL @mistercrunch

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -1379,6 +1380,12 @@ class CLIFactory(object):
              "backfill-related DAG runs and start "
              "anew with fresh, running DAG runs"),
             "store_true"),
+        'rerun_failed_tasks': Arg(
+            ("--rerun_failed_tasks",),
+            ("if set, the backfill will auto-rerun all the failed tasks for the backfill"
+             " date range instead of throwing exceptions"),
```
Comment (mistercrunch) [inline airflow/bin/cli.py]: nit: alignment seems off
nitty nit: spaces are better as last char of previous line as opposed to first char of line

File: airflow/jobs.py [hunk 2]
```diff
@@ -2241,20 +2262,40 @@ def _process_backfill_task_instances(self,
...
+                    # in case max concurrency has been reached at task runtime
+                    elif ti.state == State.NONE:
+                        self.log.warning(
+                            "FIXME: task instance {} state was set to None "
+                            "externally. This should not happen"
+                        )
+                        ti.set_state(State.SCHEDULED, session=session)
+                    if self.rerun_failed_tasks:
+                        # Rerun failed tasks or upstreamed failed tasks
+                        if ti.state == State.FAILED or ti.state == State.UPSTREAM_FAILED:
```
Comment (mistercrunch) [inline airflow/jobs.py]: `if ti.state in (State.FAILED, State.UPSTREAM_FAILED):`

File: airflow/jobs.py [hunk 3]
```diff
@@ -2241,20 +2262,40 @@ def _process_backfill_task_instances(self,
...
+                            self.log.error("Task instance %s failed "
```
Comment (mistercrunch) [inline airflow/jobs.py]: `self.log.error("Task instance %s [%s] ", ti, ti.state)`

File: airflow/jobs.py [hunk 4]
```diff
@@ -2241,20 +2262,40 @@ def _process_backfill_task_instances(self,
...
+                                           "or upstream failed", ti)
+                            if key in ti_status.running:
+                                ti_status.running.pop(key)
+                            # Reset the failed task in backfill to scheduled state
+                            ti.set_state(State.SCHEDULED, session=session)
+                    else:
+                        # Default behaviour which works for subdag.
+                        if ti.state == State.FAILED:
```
Comment (mistercrunch) [inline airflow/jobs.py]: Make just one block of code using `in (State.FAILED, State.UPSTREAM_FAILED)`, parameterize the logging with `ti.state`

File: airflow/bin/cli.py (hunk 1 above)
Comment (feng-tao) [reply airflow/bin/cli.py]: thanks. updated.

File: airflow/jobs.py (hunk 2 above)
Comment (feng-tao) [reply airflow/jobs.py]: thanks. updated.

File: airflow/jobs.py (hunk 3 above)
Comment (feng-tao) [reply airflow/jobs.py]: thanks. updated.

File: airflow/jobs.py (hunk 4 above)
Comment (feng-tao) [reply airflow/jobs.py]: thanks. updated.

**Comment by feng-tao** [comment]
https://travis-ci.com/feng-tao/incubator-airflow/builds/75714775 shows it passed the ci

**Comment by mistercrunch** [comment]
LGTM

**Comment by mistercrunch** [comment]
The CI link points to a different commit though, I'll wait for travis to clear this commit and merge.

**Comment by feng-tao** [comment]
Thanks Max @mistercrunch , it seems the CI is not very stable(fail on certain ORM, but succeed on the others). I trigger the CI again and hopefully it could clear this time.

**Pull request closed without merging**